File: src/basic_deep_learning/functionality/matrix.py
"""Module containing the Matrix class."""
from typing import Self
import random
import matrix_ops

class Matrix:
    def __init__(self,
                matrix: list[list[int|float]]) -> None:
        """Implementation of matrices as a list of lists; each inner list representing a row.
        
        Example: The following matrix M=
        
        1 2\n
        3 4
        
        is implemented as "M=Matrix([[1,2],[3,4]])".
        
        If the user passes to the matrix argument a list of uneven rows, the constructor automatically fills the empty places with 0's."""
        if not matrix:
            raise TypeError("Cannot pass an empty matrix.")
        pmax = max(len(matrix[i]) for i in range(len(matrix)))
        if pmax < 1:
            raise TypeError("Cannot pass an empty matrix.")   
        self.format = (len(matrix), pmax) #The format of the matrix in the form of a (number of rows, numbers of columns) tuple.
        self.matrix = [row + [0] * (pmax - len(row)) for row in matrix] #The list of lists representing the matrix, where the empty spaces are filled with 0's.

    def __validate_indices(self,
                           i: int|None = None,
                           j: int|None = None) -> None: #Private method.
        """If at least one index is passesd, the method will raise an error if that index is within the appropriate range."""
        if i is not None and i not in range(1,self.format[0]+1):
            raise IndexError(f"Index {i} is out of the expected range [1,{self.format[0]}].")
        if j is not None and j not in range(1,self.format[1]+1):
            raise IndexError(f"Index {j} is out of the expected range [1,{self.format[1]}].")
    
    def get_entry(self,
                  i: int,
                  j: int) -> int|float:
        """Returns the entry at the i-th row and the j-th column of the matrix.
        
        The indexation of the matrix's rows and columns are according to the mathematical convention, ie. starts at 1.
        
        Raises an error if either one of the indices is out of range."""
        self.__validate_indices(i, j)
        return self.matrix[i-1][j-1]
    
    def set_entry(self,
                  value: int|float,
                  i: int,
                  j: int) -> None:
        """Takes a certain floating-point value and sets it as the (i, j) entry.
        
        The indexation of the matrix's rows and columns are according to the mathematical convention, ie. starts at 1.
        
        Raises an error if either one of the indices is out of range."""
        self.__validate_indices(i, j)
        self.matrix[i-1][j-1] = value
    
    def get_row(self,
                i: int) -> list[int|float]:
        """Returns the i-th row.
        
        The indexation of the matrix's rows and columns are according to the mathematical convention, ie. starts at 1.
        
        Raises an error if the index is out of range."""
        self.__validate_indices(i=i)
        return self.matrix[i-1]
    
    def get_column(self,
                j: int) -> list[int|float]:
        """Returns the j-th column.
        
        The indexation of the matrix's rows and columns are according to the mathematical convention, ie. starts at 1.
        
        Raises an error if the index is out of range."""
        self.__validate_indices(j=j)
        return [row[j - 1] for row in self.matrix]
    
    @classmethod
    def zero(cls,
            n: int,
            p: int) -> Self:
        """Takes a format (number of rows, number of columns) and returns a matrix with that format and whose entries are all zeros."""
        return cls([[0 for j in range(p)] for i in range(n)])
    
    def T(self) -> Self:
        """Returns the transposed matrix."""
        transposed = [[self.matrix[j][i] for j in range(self.format[0])] for i in range(self.format[1])]
        return Matrix(transposed)

    def __add__(self, B: Self) -> Self:
        """Overloads the + operator to Matrix objects.
        
        If M and N are two Matrix instances, "M+N" returns a matrix instance representing their sum.
        
        Raises an error if the formats of the matrices do not match."""
        if self.format != B.format:
            raise TypeError("Cannot add two matrices of different formats.")
        A_cpp = matrix_ops.Matrix(self.matrix)
        B_cpp = matrix_ops.Matrix(B.matrix)  # Changed from B.format to B.matrix
        C_cpp = matrix_ops.Matrix.sum(A_cpp, B_cpp)

        return Matrix(C_cpp.get_matrix())

    def __sub__(self, B: Self) -> Self:
        """Overloads the - operator to Matrix objects.
        
        If M and N are two Matrix instances, "M-N" returns a matrix instance representing their difference.
        
        Raises an error if the formats of the matrices do not match."""
        if self.format != B.format:
            raise TypeError("Cannot subtraact two matrices of different formats.")
        A_cpp = matrix_ops.Matrix(self.matrix)
        B_cpp = matrix_ops.Matrix(B.matrix)  # Changed from B.format to B.matrix
        minusB_cpp = matrix_ops.Matrix.scale(B_cpp, -1)
        C_cpp = matrix_ops.Matrix.sum(A_cpp, minusB_cpp)

        return Matrix(C_cpp.get_matrix())
        
    def __matmul__(self, B: Self) -> Self:
        """Overloads the @ operator to Matrix objects.
        
        If M and N are two Matrix instances, "M @ N" returns a matrix instance representing their element-wise product.
        
        Raises an error if the formats of the matrices do not match."""
        if self.format != B.format:
            raise TypeError("Cannot multiply component-wise two matrices of different formats.")
        A_cpp = matrix_ops.Matrix(self.matrix)
        B_cpp = matrix_ops.Matrix(B.matrix)
        C_cpp = matrix_ops.Matrix.cwise_prod(A_cpp, B_cpp)

        return Matrix(C_cpp.get_matrix())
    
    def __mul__(self, B: Self) -> Self:
        """Overloads the * operator to Matrix objects.
        
        If M and N are two Matrix instances, "M * N" returns a matrix instance representing their product.
        
        Raises an error if the number of columns of the first does not match the number of rows of the second."""
        if self.format[1] != B.format[0]:
            raise TypeError(f"Invalid matrix formats ({self.format[1]}≠{B.format[0]}).")
        A_cpp = matrix_ops.Matrix(self.matrix)
        B_cpp = matrix_ops.Matrix(B.matrix)
        C_cpp = matrix_ops.Matrix.product(A_cpp, B_cpp)
        return Matrix(C_cpp.get_matrix())
    
    def __rmul__(self, scalar: int|float) -> Self:
        """Overloads the * operator to allow multiplication of a Matrix instance by a scalar on the left.
        
        If M is a Matrix instance and c is a scalar, "c * M" returns a matrix instance representing their product."""
        A_cpp = matrix_ops.Matrix(self.matrix)
        scaledA_cpp = matrix_ops.Matrix.scale(A_cpp, scalar)
        return Matrix(scaledA_cpp.get_matrix())
    
    def __truediv__(self, scalar):
        # Add element-wise division by scalar
        if not isinstance(scalar, (int, float)):
            raise TypeError("Can only divide by scalar")
        A_cpp = matrix_ops.Matrix(self.matrix)
        scaledA_cpp = matrix_ops.Matrix.scale(A_cpp, 1/scalar)
        return Matrix(scaledA_cpp.get_matrix())
    
    def __eq__(self, B: Self) -> bool:
        """Overloads the == operator."""
        if self.matrix == B.matrix:
            return True
        else:
            return False

    def __str__(self) -> str:
        n = self.format[0]
        s = ''
        for i in range(n-1):
            s += f'\t{self.matrix[i]},\n'
        s += f'\t{self.matrix[n-1]}'
        return f'matrix([\n{s}\n])'
    
    @classmethod
    def randomize(cls,
                  n: int,
                  p: int,
                  min_value: int|float,
                  max_value: int|float) -> Self:
        """Takes a format (n, p) and a range of values to create a random Matrix instance."""
        matrix_instance = cls.zero(n, p)
        for i in range(1, n+1):
            for j in range(1, p+1):
                t = random.random()
                matrix_instance.set_entry(min_value*(1-t) + t*max_value, i, j)
        return matrix_instance

File: src/cpp/Matrix.cpp
#include "Matrix.h"
#include <vector>
#include <ctime>
#include <cstdlib>
#include <algorithm>
using namespace std;

double dot(vector<double> u, vector<double> v){
    double dot_product = 0;
    for(int i = 0; i < u.size(); i++){
        dot_product += u[i]*v[i];
    }
    return dot_product;
}

Matrix::Matrix(vector<vector<double>> inpt_matrix){
    int n = inpt_matrix.size();

    int pmax = 0;
    for(int i = 0; i < n; i++){
        if(inpt_matrix[i].size() > pmax) {
            pmax = inpt_matrix[i].size();
        }
    }
    
    format = {n, pmax};
    matrix = inpt_matrix;
}

double Matrix::get_entry(int i, int j) const {
    return matrix[i-1][j-1];
}

void Matrix::set_entry(double val, int i, int j){
    matrix[i-1][j-1] = val;
}

vector<double> Matrix::get_row(int i) const {
    return matrix[i-1];
}

vector<double> Matrix::get_column(int j) const {
    vector<double> column;
    for(vector<double> row : matrix){
        column.push_back(row[j-1]);
    }
    return column;
}

Matrix Matrix::zero(int n, int p){
    vector<vector<double>> M;
    for(int i = 0; i < n; i++){
        vector<double> row;
        for(int j = 0; j < p; j++){
            row.push_back(0);
        }
        M.push_back(row);
    }
    return Matrix(M);
}

void Matrix::T(){
    vector<vector<double>> copy_matrix = matrix;
    int original_format[2] = {format[0], format[1]};
    matrix = Matrix::zero(original_format[1],original_format[0]).matrix;
    format = {original_format[1], original_format[0]};
    for(int i = 0; i < format[0]; i++){
        for(int j = 0; j < format[1]; j++){
            matrix[i][j] = copy_matrix[j][i];
        }
    }
}

Matrix Matrix::operator+(const Matrix& B){
    vector<vector<double>> matrix_sum;
    for(int i = 0; i < format[0]; i++){
        vector<double> row;
        for(int j = 0; j < format[1]; j++){
            row.push_back(matrix[i][j] + B.matrix[i][j]);
        }
        matrix_sum.push_back(row);
    }
    return Matrix(matrix_sum);
}

Matrix Matrix::operator-(const Matrix& B){
    vector<vector<double>> matrix_diff;
    for(int i = 0; i < format[0]; i++){
        vector<double> row;
        for(int j = 0; j < format[1]; j++){
            row.push_back(matrix[i][j] - B.matrix[i][j]);
        }
        matrix_diff.push_back(row);
    }
    return Matrix(matrix_diff);
}

Matrix Matrix::operator%(const Matrix& B){
    vector<vector<double>> matrix_cwise;
    for(int i = 0; i < format[0]; i++){
        vector<double> row;
        for(int j = 0; j < format[1]; j++){
            row.push_back(matrix[i][j] * B.matrix[i][j]);
        }
        matrix_cwise.push_back(row);
    }
    return Matrix(matrix_cwise);
}

Matrix Matrix::operator^(const Matrix& B) const {
    Matrix matrix_product = Matrix::zero(format[0], B.format[1]);
    for(int i = 1; i < format[0] + 1; i++){
        for(int j = 1; j < B.format[1] + 1; j++){
            double value = dot(get_row(i), B.get_column(j));
            matrix_product.set_entry(value, i, j);
        }
    }
    return matrix_product;
}

Matrix Matrix::randomize(int n, int p, double min_value, double max_value){
    Matrix random_matrix = Matrix::zero(n, p);
    srand(time(0));
    for(int i = 1; i < n + 1 ; i++){
        for(int j = 1; j < p + 1; j++){
            double t = static_cast<double>(rand()) / RAND_MAX;
            random_matrix.set_entry(min_value * (1 - t) + t * max_value, i, j);
        }
    }
    return random_matrix;
}